lib/seahorn/Houdini.cc

```diff
@@ -0,0 +1,292 @@
+
+    //load the Horn clause database
+    auto &db = hm.getHornClauseDB ();
+    db.buildIndexes ();
+
+    workListAlgo(db, hm);
+    //getUseRuleSet(db);
+
+    return false;
+  }
+
+  void Houdini::getAnalysisUsage (AnalysisUsage &AU) const
+  {
+    AU.addRequired<HornifyModule> ();
+    AU.setPreservesAll();
+  }
+
+  void Houdini::printDB (const HornClauseDB &db)
+  {
+    outs () << db << "\n";
+  }
+
+  void Houdini::printHello ()
+  {
+    outs () << "Hello there.\n";
+  }
+
+  void Houdini::getUseRuleSet(HornClauseDB &db)
+  {
+	  auto &workList = db.getRules();
+	  while (!workList.empty())
+	  {
+		  HornRule r = workList.front();
+		  outs() << "rule head: " << *(bind::fname(r.head())) << "\n";
+		  outs() << "rule body: " << *(r.body()) << "\n";
+		  outs() << "use size: " << db.use(bind::fname(r.head())).size() << "\n";
+		  workList.erase(workList.begin());
+	  }
+  }
+
+  Expr Houdini::guessCandidate(Expr fdecl)
+  {
+	ExprVector bvars;
+	ExprVector bins;   // a vector of LT expressions
+	Expr cand = NULL;
+
+	int bvar_count = 0;
+	unsigned i = 0;
+	for (i=0; i < bind::domainSz(fdecl); i++)
+	{
+		// if its type is INT
+		if (isOpX<INT_TY>(bind::domainTy(fdecl, i)))
+		{
+			// what is efac?
+			Expr bvar = bind::bvar (i, mk<INT_TY>(bind::domainTy(fdecl, i)->efac())); //the id of bvar is the same as related arg index
+			bvars.push_back(bvar);
+			bvar_count ++;
+		}
+	}
+	//outs() << "arg_count = " << i << "\n";
+	//outs() << "bvar count = " << bvar_count << "\n";
+
+	//What if there's no bvar?
+	if(bvar_count == 0)
+	{
+		cand = mk<TRUE>(fdecl->efac());
+	}
+	// if there is only one bvar, create a int constant and make an lt expr
+	else if(bvar_count == 1)
+	{
+		Expr zero = mkTerm<mpz_class> (0, fdecl->efac());
+		cand = mk<LT>(bvars[0], zero);
+	}
+	// if there are more than two bvars, make an lt expr
+	else if(bvar_count == 2)
+	{
+		Expr lt1 = mk<LT>(bvars[0], bvars[1]);
+		Expr lt2 = mk<LT>(bvars[1], bvars[0]);
+		bins.push_back(lt1);
+		bins.push_back(lt2);
+
+		cand = mknary<AND>(bins.begin(), bins.end());
+	}
+	else // bvar_count > 2
+	{
+		for(int j=0; j<bvars.size()-1; j++)
+		{
+			Expr lt = mk<LT>(bvars[j], bvars[j+1]);
+			bins.push_back(lt);
+		}
+		cand = mknary<AND>(bins.begin(), bins.end());
+	}
+	//outs() << *cand << "\n";
+
+	return cand;
+  }
+
+  #define SAT true
+  #define UNSAT false
+
+  void Houdini::workListAlgo(HornClauseDB &db, HornifyModule &hm)
+  {
+	  //auto &workList = db.getRules();
+	  HornClauseDB::RuleVector workList;
+
+	  for (HornClauseDB::RuleVector::iterator it = db.getRules().begin(); it != db.getRules().end(); ++it)
+	  {
+		  HornRule r = *it;
+		  workList.push_back(r);
+	  }
+
+	  while (!workList.empty())
+	  {
+		  outs() << "[WORKLIST SIZE]: " << workList.size() << "\n";
+		  HornRule r = workList.front();
+
+		  outs() << "[RULE HEAD]: " << *(r.head()) << "\n";
+		  outs() << "[RULE BODY]: " << *(r.body()) << "\n";
+
+		  //generate candidate for rule head
```

Comment:
Fixed the design of guessCandidate(), now it is called only once at the beginning to construct a global map from predicates to candidates.